src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/authorization/ReferralAccessChecker.kt

```diff
@@ -0,0 +1,29 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.authorization
+
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.AuthUser
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.Referral
+
+class ReferralAccessChecker(
+  private val userTypeChecker: UserTypeChecker,
+  private val eligibleProviderMapper: EligibleProviderMapper,
+  private val userAccessScopeMapper: UserAccessScopeMapper,
+) {
+  fun forServiceProviderUser(referral: Referral, user: AuthUser): Boolean {
+    if (!userTypeChecker.isServiceProviderUser(user)) {
+      return false
+    }
+
+    val userScope = userAccessScopeMapper.forServiceProviderUser(user)
+    val eligibleServiceProviders = eligibleProviderMapper.fromReferral(referral)
+
+    if (!eligibleServiceProviders.contains(userScope.serviceProvider)) {
```

Comment:
Kotlin ignorance question: is object equality here well defined enough for the `contains` check?

Reply:
i made a comment about this later. it does work, but i want to dig into exactly how.

Reply:
made a ticket for this